### PR TITLE
fix: add support for deleting nested fields that contain periods

### LINKED ISSUE #288: Firestore - Inconsistent behavior in key addressing
OS Type: Its Java - version 11
Firestore Java SDK version (1.33.0, but I can see that the problem still exists in the code on master)

Using field keys with dots (.), i.e. `my.key: "My value"` is allowed, albeit a bit frowned upon, has inconsistent behavior when creating compared to deleting. It is perfectly fine to create a key with dots but cannot be deleted (unless it exists on the top level of the document).

```json
{
   "my.key": true
}
```
can be removed, but:
```json
{
   "my": {
      "specialized.key": true
   }
}
```
cannot.

That's not all! Using the JavaScript SDK for Firestore, deleting is not a problem.

It is always debatable whether or not to use dots in keys, but I think the inconsistency is something that has to be fixed.

#### Steps to reproduce

1. Create a document like the one above (with specialized.key)
2. Run one of the code blocks below
```java
var updates = Map.of("my.specialized.key", FieldValue.delete());
firestore.collection("myCollection").document("myDocument").update(updates)
```
or
```java
var path = new FieldPath("my", "specialized.key");
firestore.collection("myCollection").document("myDocument").update(path, FieldValue.delete()) // <-- Very similar to JavaScript SDK
```

The first code block will succeed without any change, but the second will result in a failed Precondition.

#### Code example

See steps to reproduce

#### Stack trace

Will fail right here: https://github.com/googleapis/java-firestore/blob/80087a3ee7d1d79c099573aa5b9fb618fdff9cff/google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java#L101

If you don't have time to fix this issue(? don't know if you regard this as an issue), I can definitely contribute!

Tell me if anything is unclear!

### PR DESCRIPTION
**Pull request by suraj-qlogic**
Fixes #288

### CODE REVIEW

**Comment by suraj-qlogic** [comment]
@schmidt-sebastian ,all the changes have been addressed PTAL

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java [hunk 1]
```diff
@@ -160,7 +160,8 @@ static Value encodeValue(
       Map<String, Object> map = (Map<String, Object>) sanitizedObject;
 
       for (Map.Entry<String, Object> entry : map.entrySet()) {
-        Value encodedValue = encodeValue(path.append(entry.getKey()), entry.getValue(), options);
+        Value encodedValue =
+            encodeValue(path.append(entry.getKey(), false), entry.getValue(), options);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java:164]:
```suggestion
            encodeValue(path.append(entry.getKey(), /* splitPath= */ false), entry.getValue(), options);
```

I generally prefer these inline comments for non-descriptive values.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java [hunk 2]
```diff
@@ -1069,4 +1069,17 @@ public void updateIndividualPojo() throws ExecutionException, InterruptedExcepti
...
+  @Test
+  public void deleteNestedFieldUsingFieldPath() throws Exception {
+    doReturn(SINGLE_WRITE_COMMIT_RESPONSE)
+        .when(firestoreMock)
+        .sendRequest(
+            commitCapture.capture(), Matchers.<UnaryCallable<CommitRequest, CommitResponse>>any());
+    FieldPath path = FieldPath.of("a.b", "c.d");
+    documentReference.update(path, FieldValue.delete()).get();
+    CommitRequest expectedCommit =
+        commit(update(new HashMap<String, Value>(), Arrays.asList("`a.b`.`c.d`")));
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java:1082]:
```suggestion
        commit(update(Collections.<String, Value>emptyMap(), Collections.singletonList("`a.b`.`c.d`")));
```

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java [hunk 3]
```diff
@@ -1269,4 +1271,26 @@ public void testInstanceReturnedByGetServiceCanBeUsedToDeserializeAQuery() throw
     fs.close();
     Query.fromProto(fs, proto);
   }
+
+  @Test
+  public void deleteNestedFieldUsingFieldPath() throws Exception {
+    DocumentReference documentReference = randomColl.document("doc1");
+    Map<String, Object> dotKeyMap = new HashMap<>();
+    dotKeyMap.put("a.b", SINGLE_FILED_MAP_WITH_DOT);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java:1279]:
```suggestion
    Map<String, Object> dotKeyMap =  map(""a.b"", SINGLE_FILED_MAP_WITH_DOT);
```

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java [hunk 4]
```diff
@@ -1269,4 +1271,26 @@ public void testInstanceReturnedByGetServiceCanBeUsedToDeserializeAQuery() throw
...
+    documentReference.set(dotKeyMap).get();
+    DocumentSnapshot documentSnapshots = documentReference.get().get();
+    assertEquals(SINGLE_FILED_MAP_WITH_DOT, documentSnapshots.getData().get("a.b"));
+    FieldPath path = FieldPath.of("a.b", "c.d");
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java:1283]: Add empty line above.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java [hunk 5]
```diff
@@ -1269,4 +1271,26 @@ public void testInstanceReturnedByGetServiceCanBeUsedToDeserializeAQuery() throw
...
+    documentSnapshots = documentReference.get().get();
+    assertNull(documentSnapshots.getData().get("c.d"));
+    Map<String, Object> nestedMap = new HashMap<>();
+    nestedMap.put("e", SINGLE_FILED_MAP_WITH_DOT);
+    dotKeyMap.put("a.b", nestedMap);
+    documentReference.set(dotKeyMap).get();
+    path = FieldPath.of("a.b", "e", "c.d");
+    documentReference.update(path, FieldValue.delete()).get();
+    documentSnapshots = documentReference.get().get();
+    assertNull(documentSnapshots.getData().get("c.d"));
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java:1294]: I think you can delete this part as it doesn't expand on the tests coverage.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java (hunk 1 above)
Comment (suraj-qlogic) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java]: Done

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java (hunk 2 above)
Comment (suraj-qlogic) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java]: Done

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java (hunk 3 above)
Comment (suraj-qlogic) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java]: Done

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java (hunk 4 above)
Comment (suraj-qlogic) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java:1282]: Done

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java (hunk 5 above)
Comment (suraj-qlogic) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java]: Done

**Pull request merged**